refactor(ast): generated `Serialize` impls flatten struct fields

overlookmotel opened the pull request:
In generated `impl Serialize` for AST types, flatten the fields of `Span` directly in the `serialize` method.

Before:

```rs
self.span.serialize(serde::__private::ser::FlatMapSerializer(&mut map))?;
```

After:

```rs
map.serialize_entry("start", &self.span.start)?;
map.serialize_entry("end", &self.span.end)?;
```

Serde's derive macro can't do this, because it only knows about the type it's deriving the impl for, so it has no idea what the fields of `Span` are. But `oxc_ast_tools` knows the entire AST, so it can use that information.

I doubt this will have a runtime perf impact, but it may help compile times.

Also, shorten generated code for `Serialize` impls on enums.

overlookmotel commented:
* **#8904** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/8904?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/8904?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=8904).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

Dunqing commented:
### Merge activity

* **Feb 5, 2:08 AM EST**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.
* **Feb 5, 2:08 AM EST**: A user added this pull request to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc).

overlookmotel committed:
refactor(ast): generated `Serialize` impls flatten struct fields (#8904)

In generated `impl Serialize` for AST types, flatten the fields of `Span` directly in the `serialize` method.

Before:

```rs
self.span.serialize(serde::__private::ser::FlatMapSerializer(&mut map))?;
```

After:

```rs
map.serialize_entry("start", &self.span.start)?;
map.serialize_entry("end", &self.span.end)?;
```

Serde's derive macro can't do this, because it only knows about the type it's deriving the impl for, so it has no idea what the fields of `Span` are. But `oxc_ast_tools` knows the entire AST, so it can use that information.

I doubt this will have a runtime perf impact, but it may help compile times.

Also, shorten generated code for `Serialize` impls on enums.
